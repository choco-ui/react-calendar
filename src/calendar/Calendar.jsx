import React, { useEffect, useState } from 'react';
import { formatNumber, getDateByYearMonth, getWeeksByFirstDay } from '../util.js';
import './Calendar.less';

const WEEK_NAMES = [ '日', '一', '二', '三', '四', '五', '六' ];
const LINES = [ 1, 2, 3, 4, 5, 6 ];
const MONTH_DAYS = [ 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 ];


const Calendar = () => {

	const [ year, setYear ] = useState(0);
	const [ month, setMonth ] = useState(0);
	const [ currentDate ] = useState(new Date());
	const monthDays = MONTH_DAYS[month];
	const weekDay = getWeeksByFirstDay(year, month);

	const getDayText = (line, weekIndex, weekDay, monthDays) => {
		const number = line * 7 + weekIndex - weekDay + 1;
		return number <= 0 || number > monthDays ? '<span>&nbsp</span>' : number;
	};

	const isToday = (line, weekIndex, weekDay, monthDays) => {
		const day = getDayText(line, weekIndex, weekDay, monthDays);
		const date = new Date();
		const todayYear = date.getFullYear();
		const todayMonth = date.getMonth();
		const todayDay = date.getDate();

		return year === todayYear && month === todayMonth && day === todayDay;
	};

	const setCurrentYearMonth = (date) => {
		setYear(date.getFullYear());
		setMonth(date.getMonth());
	};

	const monthChange = (monthChanged) => {
		const monthAfter = month + monthChanged;
		const date = getDateByYearMonth(year, monthAfter);
		setCurrentYearMonth(date);
	};

	useEffect(() => {
		setCurrentYearMonth(currentDate);
	}, []);

	return (
		<table className="table">
			<caption>
				<div className="caption-header">
					<span className="arrow prev" onClick={ () => monthChange(-1) }>&#60;</span>
					<span>{ year } - { formatNumber(month) }</span>
					<span className="arrow next" onClick={ () => monthChange(1) }>&gt;</span>
				</div>
			</caption>
			<thead>
			<tr>
				{
					WEEK_NAMES.map((week, key) => {
						return <td key={ key }>{ week }</td>;
					})
				}
			</tr>
			</thead>
			<tbody>
			{
				LINES.map((l, key) => {
					return <tr key={ key }>
						{
							WEEK_NAMES.map((week, index) => {
								const fontColor = isToday(key, index, weekDay, monthDays) ? 'red' : '#000';
								return (
									<td key={ index } style={ { 'color': fontColor } }>
										{ getDayText(key, index, weekDay, monthDays) }
									</td>
								);
							})
						}
					</tr>;
				})
			}
			</tbody>
		</table>
	);
};

export default Calendar;
