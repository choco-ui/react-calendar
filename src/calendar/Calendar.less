.table {
  border-collapse: collapse;
  border-spacing: 0;
}

.table {
  td {
    border: 1px solid #ddd;
    padding: 10px;
  }

  .caption-header {
    border-top: 1px solid #ddd;
    border-right: 1px solid #ddd;
    border-left: 1px solid #ddd;
    padding: 10px;
    display: flex;
    justify-content: space-between;

    .arrow {
      cursor: pointer;
    }
  }
}
