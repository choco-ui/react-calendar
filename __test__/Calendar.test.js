import React from 'react';
import renderer from 'react-test-renderer';
import Calendar from '../src/calendar/Calendar';

describe('Calendar component', () => {
	it('snapshot', () => {
		const constant = renderer.create(<Calendar />);
		expect(constant).toMatchSnapshot();
	});
});
