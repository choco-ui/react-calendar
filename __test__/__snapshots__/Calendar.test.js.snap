// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar component snapshot 1`] = `
<table
  className="table"
>
  <caption>
    <div
      className="caption-header"
    >
      <span
        className="arrow prev"
        onClick={[Function]}
      >
        &lt;
      </span>
      <span>
        0
         - 
        01
      </span>
      <span
        className="arrow next"
        onClick={[Function]}
      >
        &gt;
      </span>
    </div>
  </caption>
  <thead>
    <tr>
      <td>
        日
      </td>
      <td>
        一
      </td>
      <td>
        二
      </td>
      <td>
        三
      </td>
      <td>
        四
      </td>
      <td>
        五
      </td>
      <td>
        六
      </td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        1
      </td>
    </tr>
    <tr>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        2
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        3
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        4
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        5
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        6
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        7
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        8
      </td>
    </tr>
    <tr>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        9
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        10
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        11
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        12
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        13
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        14
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        15
      </td>
    </tr>
    <tr>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        16
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        17
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        18
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        19
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        20
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        21
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        22
      </td>
    </tr>
    <tr>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        23
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        24
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        25
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        26
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        27
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        28
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        29
      </td>
    </tr>
    <tr>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        30
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        31
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
      <td
        style={
          Object {
            "color": "#000",
          }
        }
      >
        &lt;span&gt;&nbsp;&lt;/span&gt;
      </td>
    </tr>
  </tbody>
</table>
`;
